### PR TITLE
policy: Add selectorcache cardinality metrics

### PR DESCRIPTION
**Pull request by joestringer**
Add metrics to track the maximum cardinality of selectors actively used
to select peers in the cluster. This can be useful as a hint as to the
use of policy rules that are overly permissive, leading to high policy
map usage.

Example with some sample policies + usage:

```
  # cilium-dbg policy selectors
  SELECTOR                                                                                                                                                                       LABELS                              USERS   IDENTITIES
  &LabelSelector{...}   default/from-world-to-role-public   2       63484
  &LabelSelector{...}   default/fqdn                        1       39546
  &LabelSelector{...}   default/rule1                       1       4992
                                                                    39546
                                                                    63484
  &LabelSelector{...}   default/fqdn                        3       45238
  &LabelSelector{...}   default/fqdn                        1
  &LabelSelector{...}   default/from-world-to-role-public   1       9
                                                                    16777217
  &LabelSelector{...}   default/from-world-to-role-public   1       10
  &LabelSelector{...}   default/from-world-to-role-public   1       2
  api.github.com ...    default/fqdn                        1       16777217
  # cilium-dbg metrics list | grep selector
  cilium_fqdn_selectors                                     1.000000
  cilium_policy_selector_match_count_max   class="cluster"  0.000000
  cilium_policy_selector_match_count_max   class="fqdn"     1.000000
  cilium_policy_selector_match_count_max   class="other"    3.000000
  cilium_policy_selector_match_count_max   class="world"    2.000000
```

The cilium_fqdn_selectors metric counts the total number of fqdn
selectors altogether, but the cilium_policy_selector_match_count_max is
the maximum number of identities selected by any selector in the
specified class.

For now to avoid high cardinality for the metric itself, there are four "classes":
- World: Equivalent to entity selector on world
- Cluster: Equivalent to entity selector on the cluster
- FQDN: Any selector that is based on an FQDNs statement
- Other: All other selectors.

"All" is notably omitted, because the Cilium dataplane already optimizes
such statements to minimize cardinality when encoding the identities into
policy maps.

Review commit by commit.

```
Add new metric "cilium_policy_selector_match_count_max" which counts the maximum number of identities selected by a selector of various different selector classes.
```

### CODE REVIEW

**Review by christarazi — APPROVED** [review]
Nice!

File: pkg/policy/selectorcache.go [hunk 1]
```diff
@@ -136,6 +137,26 @@ func (sc *SelectorCache) GetModel() models.SelectorCache {
 	return selCacheMdl
 }
 
+func (sc *SelectorCache) Stats() selectorStats {
+	result := newSelectorStats()
+
+	sc.mutex.RLock()
+	defer sc.mutex.RUnlock()
```
Comment (joestringer) [inline pkg/policy/selectorcache.go:145]: I don't have a good sense for the impact of this. It's just the readlock, but how often does Prometheus call the `Collect()` function on a `Collector`? cc @cilium/metrics

**Comment by squeed** [comment]
So, as of v1.16 we also use the SelectorCache to determine subject pods. I could see a case where there is a simple broad-based policy (e.g. intercept all DNS) that would export a correspondingly large metric, but it wouldn’t be *interesting*.

Would it make sense to track whether the selector owners are peer vs. subject, and only export selectors with peer owners?

**Comment by joestringer** [comment]
@squeed ah good catch, yes I did notice subject selectors were highlighted here and I hadn't thought through how something like a clusterwide DNS policy could end up flagging false positives with this. It makes sense to me to differentiate the subject vs. peer selectors, will just have to figure out a clean solution.

File: pkg/policy/selectorcache.go (hunk 1 above)
Comment (chancez) [reply pkg/policy/selectorcache.go:145]: It occurs on each scrape, so it depends on your Prometheus scrape interval.

**Comment by joestringer** [comment]
Two new commits on top of the previous solution consolidate the testing code for dummy selectorcache users, and implement filtering to avoid reporting subject selectors in the new metric.

**Review by ianvernon — APPROVED** [review]
Nice work :)

**Review by jrajahalme — APPROVED** [review]

**Review by squeed — APPROVED** [review]

**Review by qmonnet — APPROVED** [review]

**Pull request merged**